docs/clients/fluentd/README.md

```diff
@@ -0,0 +1,209 @@
+# Fluentd
+
+Loki has a [Fluentd](https://www.fluentd.org/) output plugin called
+`fluent-plugin-grafana-loki` that enables shipping logs to a private Loki
+instance or [Grafana Cloud](https://grafana.com/oss/loki).
+
+## Installation
+
+```
+$ gem install fluent-plugin-grafana-loki
+```
+
+## Usage
+In your Fluentd configuration, use `@type loki`. Additional configuration is optional, default values would look like this:
+```
+<match **>
+  @type loki
+  url "https://logs-us-west1.grafana.net"
+  username "#{ENV['LOKI_USERNAME']}"
+  password "#{ENV['LOKI_PASSWORD']}"
+  extra_labels {"env":"dev"}
+  flush_interval 10s
+  flush_at_shutdown true
+  buffer_chunk_limit 1m
+</match>
+```
+
+### Using labels
+
+Simple label from top level attribute
+```
+<match mytag>
+  @type loki
+  # ...
+  <label>
+    fluentd_worker
+  </label>
+  # ...
+</match>
+```
+
+You can rewrite the label keys as well as the following
+
+```
+<match mytag>
+  @type loki
+  # ...
+  <label>
+    worker fluentd_worker
+  </label>
+  # ...
+</match>
+```
+
+You can use record accessor syntax for nested field. https://docs.fluentd.org/plugin-helper-overview/api-plugin-helper-record_accessor#syntax
+
+```
+<match mytag>
+  @type loki
+  # ...
+  <label>
+    container $.kubernetes.container
+  </label>
+  # ...
+</match>
+```
+
+### Extracting Kubernetes labels
+
+As Kubernetes labels are a list of nested key-value pairs there is a separate option to extract them.
+Note that special characters like "`. - /`" will be overwritten with `_`.
+Use with the `remove_keys kubernetes` option to eliminate metadata from the log.
+```
+<match mytag>
+  @type loki
+  # ...
+  extract_kubernetes_labels true
+  remove_keys kubernetes
+  <label>
+    container $.kubernetes.container
+  </label>
+  # ...
+</match>
+```
+
+### Multi-worker usage
+
+Loki doesn't currently support out-of-order inserts - if you try to insert a log entry an earlier timestamp after a log entry with with identical labels but a later timestamp, the insert will fail with `HTTP status code: 500, message: rpc error: code = Unknown desc = Entry out of order`. Therefore, in order to use this plugin in a multi worker Fluentd setup, you'll need to include the worker ID in the labels.
```

Comment:
```suggestion
Loki doesn't currently support out-of-order inserts - if you try to insert a log entry an earlier timestamp after a log entry with with identical labels but a later timestamp, the insert will fail with `HTTP status code: 500, message: rpc error: code = Unknown desc = Entry out of order`. Therefore, in order to use this plugin in a multi worker Fluentd setup, you'll need to include the worker ID in the labels or otherwise ensure log streams are always sent to the same worker.
```

Reply:
Not sure if this is possible (I haven't used fluentd), but wanted to put in this small disclaimer

Reply:
What about this https://github.com/grafana/loki/pull/1634/commits/3f20282e7f505aa6ecb28ec81d3efaded7a34101 ?